### PR TITLE
ApiKey additions

### PR DESCRIPTION
**Pull request by rukai**
This PR introduces 2 new pieces of functionality around ApiKey
These were required to implement a unit test in my project that compares the currently supported versions of the project against the supported versions of the kafka-protocol crate.

The additions are:
* Add `ApiKey::iterate_all()` function, it returns an iterator that iterates once over every ApiKey variant
* Add `ApiKey::valid_versions(&self)` method, it returns the valid versions that can be used with this ApiKey variant.

I'm happy to split the PR in two if one of these is easy to land and the other is not.

### CODE REVIEW

File: protocol_codegen/src/generate_messages.rs [hunk 1]
```diff
@@ -179,6 +184,40 @@ pub fn run(messages_module_dir: &str, mut input_file_paths: Vec<PathBuf>) -> Res
     }
     writeln!(m, "        }}")?;
     writeln!(m, "    }}")?;
+
+    writeln!(
+        m,
+        "    /// Returns the valid versions that can be used with this ApiKey"
```
Comment (tychedelia) [inline protocol_codegen/src/generate_messages.rs:190]: This seems immensely useful!

File: src/messages.rs [hunk 2]
```diff
@@ -1324,6 +1325,92 @@ impl ApiKey {
...
+            ApiKey::GetTelemetrySubscriptionsKey => VersionRange { min: 0, max: 0 },
+            ApiKey::PushTelemetryKey => VersionRange { min: 0, max: 0 },
+            ApiKey::AssignReplicasToDirsKey => VersionRange { min: 0, max: 0 },
+            ApiKey::ListClientMetricsResourcesKey => VersionRange { min: 0, max: 0 },
+            ApiKey::DescribeTopicPartitionsKey => VersionRange { min: 0, max: 0 },
+        }
+    }
+
+    /// Iterate through every ApiKey variant in the order of the internal code.
+    pub fn iterate_all() -> impl Iterator<Item = ApiKey> {
```
Comment (tychedelia) [inline src/messages.rs:1411]: Nit: Any reason not just to call this `iter`? It seems like that's what [strum](https://docs.rs/strum_macros/latest/strum_macros/derive.EnumIter.html) does so is potentially idiomatic for enum variants.

**Review by tychedelia — APPROVED** [review]

**Comment by tychedelia** [comment]
Oy, apologies, I think I may have merged these out of order and caused conflicts here. Messages needs to be regenerated.

File: src/messages.rs (hunk 2 above)
Comment (rukai) [reply src/messages.rs:1411]: yeah, good point, I'll rename it

**Comment by tychedelia** [comment]
@rukai lmk when you think we should do a release.

**Comment by rukai** [comment]
I'm thinking now is a good time, I don't have any other changes in mind atm

**Pull request merged**